flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/backpressure/BackPressureRequestCoordinator.java

```diff
@@ -0,0 +1,365 @@
+	 */
+	@VisibleForTesting
+	void collectTaskBackPressureStat(
+			int requestId,
+			ExecutionAttemptID executionId,
+			double taskBackPressureRatio) {
+
+		synchronized (lock) {
+			if (isShutDown) {
+				return;
+			}
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Collecting back pressure request {} result of task {}.", requestId, executionId);
+			}
+
+			PendingBackPressureRequest pending = pendingRequests.get(requestId);
+
+			if (pending != null) {
+				pending.collectBackPressureStats(executionId, taskBackPressureRatio);
+
+				// Publish the request result
+				if (pending.isComplete()) {
+					pendingRequests.remove(requestId);
+					rememberRecentRequestId(requestId);
+
+					pending.completePromiseAndDiscard();
+				}
+			} else if (recentPendingRequests.contains(requestId)) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Received late back pressure request {} result of task {}.",
+							requestId, executionId);
+				}
+			} else {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Unknown request ID " + requestId);
+				}
+			}
+		}
+	}
+
+	private void rememberRecentRequestId(int requestId) {
+		if (recentPendingRequests.size() >= NUM_GHOST_REQUEST_IDS) {
+			recentPendingRequests.removeFirst();
+		}
+		recentPendingRequests.addLast(requestId);
+	}
+
+	@VisibleForTesting
+	int getNumberOfPendingRequests() {
+		synchronized (lock) {
+			return pendingRequests.size();
+		}
+	}
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * A pending task back pressure request, which collects task back pressure
+	 * ratio and owns a {@link BackPressureStats} promise.
+	 *
+	 * <p>Access pending request in lock scope.
+	 */
+	private static class PendingBackPressureRequest {
+
+		private final int requestId;
+		private final long startTime;
+		private final Set<ExecutionAttemptID> pendingTasks;
+		private final Map<ExecutionAttemptID, Double> backPressureRatios;
+		private final CompletableFuture<BackPressureStats> backPressureStatsFuture;
+
+		private boolean isDiscarded;
+
+		PendingBackPressureRequest(
+				int requestId,
+				ExecutionAttemptID[] tasksToCollect) {
+
+			this.requestId = requestId;
+			this.startTime = System.currentTimeMillis();
+			this.pendingTasks = new HashSet<>(Arrays.asList(tasksToCollect));
+			this.backPressureRatios = Maps.newHashMapWithExpectedSize(tasksToCollect.length);
+			this.backPressureStatsFuture = new CompletableFuture<>();
+		}
+
+		boolean isComplete() {
+			if (isDiscarded) {
+				throw new IllegalStateException("Discarded");
+			}
+
+			return pendingTasks.isEmpty();
+		}
+
+		void discard(Throwable cause) {
+			if (!isDiscarded) {
+				pendingTasks.clear();
+				backPressureRatios.clear();
+
+				backPressureStatsFuture.completeExceptionally(new RuntimeException("Discarded", cause));
+
+				isDiscarded = true;
+			}
+		}
+
+		void collectBackPressureStats(ExecutionAttemptID executionId, double backPressureRatio) {
+			if (isDiscarded) {
```

Comment:
It is better to provide two check methods for reuse in this class, then this method would be :

```
checkDiscarded();
checkCompleted();

if (pendingTasks.remove(executionId)) {
	backPressureRatios.put(executionId, backPressureRatio);
} else {
	throw new IllegalArgumentException("Unknown task " + executionId);
}
```